### ISSUE TITLE
PR 7 — `ONNXBackend` + `TensorRTBackend`

### ISSUE THREAD

**Issue by gitttt-1234** [talmolab/sleap-nn#515]
**Parent epic:** #508
**Prerequisites:** #511, #512
**Blocks:** #518, #519

---

## Goal

Implement `ONNXBackend` and `TensorRTBackend` against the `ModelBackend` protocol from #511. These backends wrap the existing `export/wrappers/` modules (which bake normalize + peak-finding + PAF scoring into the graph) and let every `InferenceLayer` work with ONNX / TensorRT runtimes unchanged.

Deletes `sleap_nn/export/predictors/onnx.py` and `tensorrt.py` (the old runtime wrappers) — their logic moves into these backends.

## Key insight: the `does_baked_postproc` dispatch

From design doc §4.3: `InferenceLayer.postprocess()` checks `self.backend.does_baked_postproc`:
- `False` (Torch): run Python peak finding + integral refinement on raw confmaps
- `True` (ONNX/TRT): the backend already returned `{"peaks": ..., "peak_vals": ...}` — just apply coord transforms

This is what makes the same layer class work with any backend. No model-type-specific code lives in the backends themselves.

## But: "baked postproc" isn't fully True/False

Per discussion / analysis of `export/wrappers/`:

| What's baked in ONNX | Single | Centroid | TopDown | BottomUp | MC variants |
|---|:-:|:-:|:-:|:-:|:-:|
| uint8 → float normalization | ✓ | ✓ | ✓ | ✓ | ✓ |
| Input scaling | ✓ | ✓ | ✓ | ✓ | ✓ |
| Rough peak finding (argmax) | ✓ | ✓ (top-K + NMS) | ✓ | ✓ (top-K per node) | ✓ |
| **Integral refinement (sub-pixel)** | **✗** | **✗** | **✗** | **✗** | **✗** |
| Crop extraction (top-down) | — | — | ✓ (grid_sample) | — | ✓ |
| PAF line scoring (bottom-up) | — | — | — | ✓ | — |
| **PAF Hungarian grouping** | — | — | — | **✗** (Python) | — |
| Class decoding | — | — | — | — | ✗ (Python) |

**Correctness gap — locked decision (design doc §7 item 7):** ONNX wrappers today do *not* include integral refinement. Post-refactor, **Option A is mandatory**: bake grid_sample-based integral refinement into the ONNX graph so ONNX and Torch produce identical outputs.

Option B (accept the sub-pixel delta) is **rejected** — it perpetuates correctness drift and keeps duplicate code alive.

### What this PR deletes from `export/wrappers/`

Per design doc §4C, `export/wrappers/` becomes a composition layer, not a reimplementation layer. Functions that go away:

| Deleted from | Replaced with |
|---|---|
| `export/wrappers/base.py::_find_global_peaks` | `from sleap_nn.inference.ops.peaks import find_global_peaks` (now with integral refinement path exportable via #513) |
| `export/wrappers/base.py::_find_topk_peaks` | `from sleap_nn.inference.ops.peaks import find_top_k_peaks` |
| `export/wrappers/base.py::_find_topk_peaks_per_node` | `from sleap_nn.inference.ops.peaks import find_top_k_peaks_per_node` |
| `export/wrappers/base.py::_neighbor_max` | `from sleap_nn.inference.ops.peaks import neighbor_max_nms` (or replaced with `F.max_pool2d`) |
| `export/wrappers/topdown.py::_extract_crops` | `from sleap_nn.inference.ops.crops import extract_instance_crops` |
| `export/wrappers/bottomup.py::_sample_and_score_lines` | `from sleap_nn.inference.ops.paf import score_lines` |

Target deletion: ~300 lines removed from `export/wrappers/`. Each wrapper becomes a thin composition of:
1. uint8 → float normalization (stays — it's the one wrapper-specific piece)
2. Input-scale `F.interpolate` (stays — wrapper-specific shape handling)
3. Model forward
4. Calls to `ops.peaks` / `ops.coord` / `ops.crops` / `ops.paf.score_lines`

No wrapper-local peak finding logic survives.

### `ONNXBackend`

```python
# sleap_nn/inference/layers/backends/onnx_backend.py

from typing import Dict, Tuple
import attrs
import numpy as np
import onnxruntime as ort
import torch


@attrs.define
class ONNXBackend:
    """ONNX Runtime backend.

    Wraps an onnxruntime.InferenceSession produced from export/wrappers/.
    The session already bakes normalize + peak-finding into the graph.

    Uses IOBinding for zero-copy GPU when device is CUDA.
    """

    session: ort.InferenceSession = attrs.field(repr=False)
    _device: str = "cpu"
    _input_name: str = attrs.field(init=False)
    _output_names: Tuple[str, ...] = attrs.field(init=False)

    def __attrs_post_init__(self):
        self._input_name = self.session.get_inputs()[0].name
        self._output_names = tuple(o.name for o in self.session.get_outputs())

    @property
    def device(self) -> str:
        return self._device

    @property
    def does_baked_postproc(self) -> bool:
        return True

    def __call__(self, x: torch.Tensor) -> Dict[str, torch.Tensor]:
        # ONNX wrapper expects uint8 input (it bakes normalization)
        if x.dtype != torch.uint8:
            x = (x * 255).clamp(0, 255).to(torch.uint8)

        if self._device.startswith("cuda"):
            return self._run_with_iobinding(x)
        return self._run_cpu(x)

    def _run_cpu(self, x):
        x_np = x.cpu().numpy()
        results = self.session.run(self._output_names, {self._input_name: x_np})
        return {
            name: torch.from_numpy(arr)
            for name, arr in zip(self._output_names, results)
        }

    def _run_with_iobinding(self, x):
        """Zero-copy path for CUDA."""
        io_binding = self.session.io_binding()
        io_binding.bind_input(
            name=self._input_name,
            device_type="cuda",
            device_id=x.device.index or 0,
            element_type=np.uint8,
            shape=tuple(x.shape),
            buffer_ptr=x.data_ptr(),
        )
        for name in self._output_names:
            io_binding.bind_output(name=name, device_type="cuda")
        self.session.run_with_iobinding(io_binding)
        outputs = io_binding.get_outputs()
        return {
            name: torch.from_numpy(out.numpy())
            for name, out in zip(self._output_names, outputs)
        }

    def warmup(self, input_shape: Tuple[int, ...]) -> None:
        dummy = torch.zeros(input_shape, dtype=torch.uint8)
        self(dummy)

    @classmethod
    def from_export_dir(cls, export_dir: str, device: str = "cpu") -> "ONNXBackend":
        """Load ONNX model + metadata from an export directory."""
        import json
        from pathlib import Path
        export_dir = Path(export_dir)
        providers = {
            "cuda": ["CUDAExecutionProvider", "CPUExecutionProvider"],
            "cpu": ["CPUExecutionProvider"],
        }[device.split(":")[0] if ":" in device else device]
        session = ort.InferenceSession(
            str(export_dir / "model.onnx"), providers=providers,
        )
        return cls(session=session, _device=device)
```

### `TensorRTBackend`

```python
# sleap_nn/inference/layers/backends/tensorrt_backend.py

from typing import Dict, Tuple
import attrs
import torch


@attrs.define
class TensorRTBackend:
    """TensorRT backend with explicit CUDA stream.

    Requires the engine to be built against the export/wrappers/ ONNX graph.
    Provides same baked-postproc semantics as ONNXBackend.
    """

    engine: "trt.ICudaEngine" = attrs.field(repr=False)
    _device: str = "cuda:0"
    _context: "trt.IExecutionContext" = attrs.field(init=False, repr=False)
    _stream: "torch.cuda.Stream" = attrs.field(init=False, repr=False)

    @property
    def device(self) -> str:
        return self._device

    @property
    def does_baked_postproc(self) -> bool:
        return True

    def __call__(self, x: torch.Tensor) -> Dict[str, torch.Tensor]:
        ...  # Standard TRT execute_async_v3 with CUDA stream

    def warmup(self, input_shape: Tuple[int, ...]) -> None:
        dummy = torch.zeros(input_shape, dtype=torch.uint8, device=self._device)
        self(dummy)

    @classmethod
    def from_export_dir(cls, export_dir: str, device: str = "cuda:0") -> "TensorRTBackend":
        ...  # Deserialize .trt engine, create execution context
```

## Handling the bottom-up mixed case

Bottom-up `InferenceLayer.postprocess()` needs to handle this specifically:

- ONNX wrapper returns `(peaks, line_scores, candidate_mask)` — peaks and line scores are baked
- PAF Hungarian grouping (`PAFScorer.predict`) still runs in Python

The fix in `BottomUpLayer.postprocess()`:

```python
def postprocess(self, raw_out, info):
    if self.backend.does_baked_postproc:
        # ONNX path: peaks + line_scores already computed on GPU.
        # Skip find_local_peaks; jump straight to PAF grouping.
        peaks = raw_out["peaks"]
        line_scores = raw_out["line_scores"]
        candidate_mask = raw_out["candidate_mask"]
        predicted_instances = self.paf_scorer.predict_from_scores(
            peaks, line_scores, candidate_mask,
        )
    else:
        # Torch path: full postproc
        ...
```

This means `PAFScorer` needs a `predict_from_scores()` entry point that skips the line-sampling step (since the ONNX wrapper already did it). Check if this already exists in `paf_grouping.py`.

## Backend consistency tests (new regression gate)

We currently have no test that compares Torch vs ONNX outputs. #515 adds one:

```python
# tests/inference/backends/test_backend_consistency.py

@pytest.mark.parametrize("model_type", [...])
def test_torch_vs_onnx(model_type):
    """Same layer, same image, two backends — outputs must match."""
    image = load_test_image()

    torch_layer = layer_for(model_type, backend="torch")
    onnx_layer = layer_for(model_type, backend="onnx")

    torch_out = torch_layer.predict(image)
    onnx_out = onnx_layer.predict(image)

    torch.testing.assert_close(
        torch_out.pred_keypoints, onnx_out.pred_keypoints,
        rtol=1e-4, atol=1e-3,  # looser than Torch parity — see integral gap above
    )
```

## Acceptance criteria

- [ ] `ONNXBackend` with CPU + CUDA (IOBinding) paths
- [ ] `TensorRTBackend` with CUDA stream
- [ ] Both pass `isinstance(..., ModelBackend)` at runtime
- [ ] Every `InferenceLayer` subclass works with both backends — parametrized parity tests pass
- [ ] Backend consistency tests pass within 1e-3 (looser than Torch parity due to argmax-only peaks; tighten if we bake integral refinement)
- [ ] `from_export_dir()` factory works on existing exported fixtures
- [ ] `sleap_nn/export/predictors/onnx.py` and `tensorrt.py` deleted; downstream imports updated
- [ ] `ruff check` + `black --check` clean

## Files touched

**Created:**
- `sleap_nn/inference/layers/backends/onnx_backend.py`
- `sleap_nn/inference/layers/backends/tensorrt_backend.py`
- `tests/inference/backends/test_onnx_backend.py`
- `tests/inference/backends/test_tensorrt_backend.py`
- `tests/inference/backends/test_backend_consistency.py`

**Modified:**
- `sleap_nn/inference/layers/backends/loader.py` — extend to dispatch on file extension (`.onnx`, `.trt`)
- `sleap_nn/inference/layers/bottomup.py` — handle the `does_baked_postproc=True` path
- `sleap_nn/inference/ops/paf.py` — add `predict_from_scores()` entry point if missing

**Deleted:**
- `sleap_nn/export/predictors/onnx.py`
- `sleap_nn/export/predictors/tensorrt.py`
- `sleap_nn/export/predictors/__init__.py` (the whole package goes)

## Out of scope

- Deleting `sleap_nn/export/inference.py` — #519
- CLI integration — #518
- Export command (`sleap-nn export`) — unchanged; lives in `sleap_nn/export/cli.py`

**Comment by gitttt-1234** [comment]
Closed — merged as part of the inference refactor stack (epic #508).
